Fix an assert in canonicalizeFunctionArgument.

atrick opened the pull request:
Assertion failed: (succeed && "should be filtered by FindBorrowScopeUses"), function canonicalizeFunctionArgument, file CanonicalizeBorrowScope.cpp, line 798

Canonicalization for guaranteed function arguments is triggered by
SILCombine without any up-front analysis. Because the canonicalization
rewrites the function argument's copies in place, it must always
succeed.

Fix the visitBorrowScopeUses utility to be aware that it is being
invoked on a function argument and avoid bailing out.

---

To make this problem reproduce reliably, I enabled it in the CopyPropagation pass. This way, the code in question runs over all SIL. Previously, it happened to be invoked during certain SILCombine optimization and could only assert in a rare case involving unmanaged references.

atrick committed:
Fix an assert in canonicalizeFunctionArgument.

Assertion failed: (succeed && "should be filtered by
FindBorrowScopeUses"), function canonicalizeFunctionArgument, file
CanonicalizeBorrowScope.cpp, line 798

Canonicalization for guaranteed function arguments is triggered by
SILCombine without any up-front analysis. Because the canonicalization
rewrites the function argument's copies in place, it must always
succeed.

Fix the visitBorrowScopeUses utility to be aware that it is being
invoked on a function argument and avoid bailing out.

atrick commented:
@swift-ci smoke test

nate-chandler commented on test/SILOptimizer/copy_propagation_borrow.sil:
```diff
@@ -1178,3 +1179,24 @@ bb0(%0 : @guaranteed $String):
+// because it begins rewriting guaranteed function arguments without
+// checking the uses first.
+//
+// CHECK-LABEL-NOSCOPES: sil [ossa] @testEscapingGuaranteedToOwned : $@convention(method) (@guaranteed C) -> () {
```
Should this be `CHECK-NOSCOPES-LABEL`?

nate-chandler commented on lib/SILOptimizer/Utils/CanonicalizeBorrowScope.cpp:
```diff
@@ -261,8 +263,13 @@ bool CanonicalizeBorrowScope::visitBorrowScopeUses(SILValue innerValue,
+            && !isa<SILFunctionArgument>(innerValue)) {
+          return false;
+        }
+        if (!visitor.visitUse(use)) {
```
Would asserting inside the condition that `!isa<SILFunctionArgument>(innerValue)` be worthwhile?

nate-chandler reviewed (approved):
Great to be doing borrow canonicalizations for all guaranteed function arguments!

atrick committed:
Disable MandatoryCopyPropagation.

Mandatory copy propagation was primarily a stop-gap until lexcial
lifetimes were implemented. It supposedly made variables lifetimes
more consistent between -O and -Onone builds. Now that lexical
lifetimes are enabled, it is no longer needed for that purpose (and
will never satisfactorily meet that goal anyway).

Mandatory copy propagation may be enabled again later as a -Onone "
optimization. But that requires a more careful audit of the effect on
debug information.

For now, it should be disabled.

atrick commented:
@swift-ci test

atrick commented:
SwiftREPL/BreakpointSimple.test

atrick commented:
@swift-ci smoke test and merge
